go/adbc/driver/databricks/cmd_record_reader.go

```diff
@@ -0,0 +1,187 @@
+)
+type cmdReader struct {
+	refCount int64
+
+	cmdExecution  compute.CommandExecutionInterface
+
+	// Statement that this reader is associated with.
+	CommandId string
+
+	CommandResult *compute.Results
+
+	rec arrow.Record
+	err error
+	schema *arrow.Schema
+	rec_read bool
+	cancelFn context.CancelFunc	
+	
+	// Statistics
+
+	// Reader's start time.
+	startTime time.Time
+	// All the bytes received from the server.
+	BytesReceived int64
+	// Time spent waiting for the server to respond in the foreground.
+	WaitTime time.Duration
+}
+
+func DeriveSchema(dbx_schema []map[string]interface{}) *arrow.Schema {
+	fields := make([]arrow.Field, len(dbx_schema))
+	for i, col := range dbx_schema {
+		var arrowType arrow.DataType
+		switch col["type"] {
+		case "string":
+			arrowType = arrow.BinaryTypes.String
+		case "int":
+			arrowType = arrow.PrimitiveTypes.Int32
+		case "long":
+			arrowType = arrow.PrimitiveTypes.Int64
+		case "float":
+			arrowType = arrow.PrimitiveTypes.Float32
+		case "double":
+			arrowType = arrow.PrimitiveTypes.Float64
+		case "boolean":
+			arrowType = arrow.FixedWidthTypes.Boolean
+		case "timestamp":
+			arrowType = arrow.FixedWidthTypes.Timestamp_us
+		case "date":
+			arrowType = arrow.FixedWidthTypes.Date32
+		default:
+			arrowType = arrow.BinaryTypes.String
+		}
+		fields[i] = arrow.Field{
+			Name:     col["name"].(string),
+			Type:     arrowType,
+			Nullable: true,
+		}
+	}
+	return arrow.NewSchema(fields, nil)
+}
+
+func NewCommandRecordReader(
+	cmdExecution compute.CommandExecutionInterface, commandId string, result *compute.Results) (*cmdReader, error) {
+	// Convert the Databricks schema to an Arrow schema
+	schema := DeriveSchema(result.Schema)
+	r := &cmdReader{
+		refCount: 1,
+
+		cmdExecution: cmdExecution,
+
+		CommandId:     commandId,
+		CommandResult: result,
+
+		schema: schema,
+		rec:    nil,
+		err:    nil,
+		rec_read: false,
+
+		cancelFn: func() {},
+
+		startTime:     time.Now(),
+		BytesReceived: 1,
+		WaitTime:      0,
+	}
+
+	// For command execution, we need to convert the result to an Arrow record
+	if result.ResultType == compute.ResultTypeTable {
+
+		switch result.Data.(type) {
+		case []interface{}:
+			rows := result.Data.([]interface{})
+			r.rec, r.err = BuildFromRows(schema, rows)
+		default:
+			r.err = adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  fmt.Sprintf("Unexpected command result type: %T", result.Data),
+			}
+		}
+
+	} else if result.ResultType == compute.ResultTypeText {
+		// For text result, return an empty record with a single string column
+		fields := []arrow.Field{{Name: "text", Type: arrow.BinaryTypes.String, Nullable: true}}
+		schema := arrow.NewSchema(fields, nil)
+		rows := result.Data.([]interface{})
+		r.rec, r.err = BuildFromRows(schema, rows)
+	} else {
+		r.err = adbc.Error{
+			Code: adbc.StatusInternal,
+			Msg:  fmt.Sprintf("Unexpected command result type: %s", result.ResultType),
+		}
+	}
```

Comment:
This logic belongs in the reader's `Next()` method to allow the record batch to be returned *before* we spend time processing the results. That' what the reader is for: reading the results :)